### PR TITLE
feat: initial module dns-zone

### PR DESCRIPTION
**Pull request by vlahane**
## Description
Closes: https://github.com/Azure/bicep-registry-modules/issues/340 

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.
- [ ] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [ ] I have updated the examples in README with the latest module version number.

### CODE REVIEW

**Comment by vlahane** [comment]
@dciborow any idea about below error, CI failing: 
```
[31;1mNew-AzResourceGroupDeployment: [31;1m5:12:32 PM - The deployment 'main.test' failed with error(s). Showing 1 out of 1 error(s).
Status Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details. (Code: DeploymentFailed)
 - {
  "code": "Conflict",
  "message": "Resource /subscriptions/xxxx/resourceGroups/xxxxx/providers/Microsoft.Network/trafficmanagerprofiles/trafdns-cjjgmyqluxcj4/$fqdn/tmp-6rorwm3vkhz3u.trafficmanager.net is locked by record set /buckets/7/com/example/myendpoint2. Retry last operation on record set to mitigate."
} (Code:Conflict)
CorrelationId: 95af5b70-8b46-4c92-a80e-eda3eed3a493[0m
```

**Comment by shenglol** [comment]
> @dciborow Daniel Ciborowski FTE any idea about below error, CI failing:
> 
> ```
> �[31;1mNew-AzResourceGroupDeployment: �[31;1m5:12:32 PM - The deployment 'main.test' failed with error(s). Showing 1 out of 1 error(s).
> Status Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details. (Code: DeploymentFailed)
>  - {
>   "code": "Conflict",
>   "message": "Resource /subscriptions/xxxx/resourceGroups/xxxxx/providers/Microsoft.Network/trafficmanagerprofiles/trafdns-cjjgmyqluxcj4/$fqdn/tmp-6rorwm3vkhz3u.trafficmanager.net is locked by record set /buckets/7/com/example/myendpoint2. Retry last operation on record set to mitigate."
> } (Code:Conflict)
> CorrelationId: 95af5b70-8b46-4c92-a80e-eda3eed3a493�[0m
> ```

@vlahane Is the error reproducible in your own subscription?

**Comment by vlahane** [comment]
> > @dciborow Daniel Ciborowski FTE any idea about below error, CI failing:
> > ```
> > �[31;1mNew-AzResourceGroupDeployment: �[31;1m5:12:32 PM - The deployment 'main.test' failed with error(s). Showing 1 out of 1 error(s).
> > Status Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details. (Code: DeploymentFailed)
> >  - {
> >   "code": "Conflict",
> >   "message": "Resource /subscriptions/xxxx/resourceGroups/xxxxx/providers/Microsoft.Network/trafficmanagerprofiles/trafdns-cjjgmyqluxcj4/$fqdn/tmp-6rorwm3vkhz3u.trafficmanager.net is locked by record set /buckets/7/com/example/myendpoint2. Retry last operation on record set to mitigate."
> > } (Code:Conflict)
> > CorrelationId: 95af5b70-8b46-4c92-a80e-eda3eed3a493�[0m
> > ```
> 
> @vlahane Is the error reproducible in your own subscription?

@shenglol  I tried again, and it works on my subscription without any error.

**Comment by shenglol** [comment]
@dciborow Do you have any idea about the deployment error?

**Comment by dciborow** [comment]
> @dciborow Do you have any idea about the deployment error?

https://github.com/Azure/bicep-registry-modules/pull/363 will fix this.

File: modules/network/dns-zone/main.bicep [hunk 1]
```diff
@@ -0,0 +1,74 @@
+@description('Required. Specifies the resource name')
+param name string
```
Comment (dciborow) [inline modules/network/dns-zone/main.bicep:2]:
```suggestion
@description('Prefix of DNS Resource Name. This param is ignored when name is provided.')
param prefix string = 'azure-'

@description('Required. Specifies the resource name')
param name string = '${prefix}${uniqueString(resourceGroup().id, subscription().id)}'
```

File: modules/network/dns-zone/main.bicep [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  name: record.name
+  parent: dnszones
+  properties: {
+    AAAARecords: (record.aliasRecordSet) ? null : record.aaaaRecord
+    TTL: record.ttl
+    targetResource: record.aliasRecordSet ? {
+      id: record.targetResource
+    } : {}
+  }
+}]
```
Comment (dciborow) [inline modules/network/dns-zone/main.bicep:68]:
```suggestion
resource dnszones 'Microsoft.Network/dnsZones@2018-05-01' = {
  name: name
  location: 'global'
  tags: tags
  properties: {
    zoneType: zoneType
  }


  resource A 'A' = [for record in aRecords: {
    name: record.name
    properties: {
      ARecords: (record.aliasRecordSet) ? null : record.aRecord
      TTL: record.aliasRecordSet ? 0 : record.ttl
      targetResource: record.aliasRecordSet ? {
        id: record.targetResource
      } : {}
    }
  }]
  
  resource CName 'CNAME' = [for record in cnameRecords: {
    name: record.name
    properties: {
      CNAMERecord: record.aliasRecordSet ? null : {
        cname: record.cname
      }
      TTL: record.ttl
      targetResource: record.aliasRecordSet ? {
        id: record.targetResource
      } : {}
    }
  }]
  
  resource AAAA 'AAAA' = [for record in AAAARecords: {
    name: record.name
    properties: {
      AAAARecords: (record.aliasRecordSet) ? null : record.aaaaRecord
      TTL: record.ttl
      targetResource: record.aliasRecordSet ? {
        id: record.targetResource
      } : {}
    }
  }]
}
```

go with nested resource declaration when able.

Comment (vlahane) [reply modules/network/dns-zone/main.bicep]: any reasons for nested resource declaration? 
does current change I have to make it nested resource declaration?

Comment (dciborow) [reply modules/network/dns-zone/main.bicep]: the main thing I like about it is it reduced the instances of the API-Version to a single place. It also saves a few lines. I find it a little easier to read.
But not a big deal.

**Review by dciborow — APPROVED** [review]

File: modules/network/dns-zone/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+| :---------- | :----: | :---------------------------------- |
+| id          | string | Id for DNS zone                     |
+| nameServers | array  | The name servers for this DNS zone. |
+
+## Examples
+
+### Example 1
+
+```bicep
+module test01 'br/public:network/dns-zone:0.1.0' = {
```
Comment (dciborow) [inline modules/network/dns-zone/README.md:39]:
```suggestion
module test01 'br/public:network/dns-zone:1.0.1' = {
```

File: modules/network/dns-zone/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  params: {
+    name: 'myendpoint1.example.com'
+  }
+}
+```
+
+### Example 2
+
+```bicep
+module test02 'br/public:network/dns-zone:0.1.0' = {
```
Comment (dciborow) [inline modules/network/dns-zone/README.md:50]:
```suggestion
module test02 'br/public:network/dns-zone:1.0.1' = {
```

File: modules/network/dns-zone/main.bicep (hunk 2 above)
Comment (vlahane) [reply modules/network/dns-zone/main.bicep]: oh, reduced the instances of the API-Version that's a big deal, really like nested resource declaration.

**Pull request merged**